### PR TITLE
Basic Prometheus support for providing Tokio Runtime metrics

### LINKED ISSUE #125: Simple prometheus metrics
**Is your feature request related to a problem? Please describe.**
I'd like to observe some minimal metrics.

**Describe the solution you'd like**
Expose prometheus metrics 

**Describe alternatives you've considered**
Looking for a rust library to possibly plugin into.

**Additional context**
Add any other context or screenshots about the feature request here if needed.

### PR DESCRIPTION
**Pull request by pl4nty**
## Description

Implements an optional Prometheus metrics endpoint at `/metrics`. Currently provides [Tokio RuntimeMetrics](https://docs.rs/tokio/latest/tokio/runtime/struct.RuntimeMetrics.html) via [tokio-metrics-collector](https://github.com/Hanaasagi/tokio-metrics-collector), requiring `tokio_unstable`. I tried implementing task metrics for the HTTPS redirect Tokio tasks, but no metrics were produced.

## Related Issue

Fixes #125

## Motivation and Context

Many cloud-native environments (eg Kubernetes) use the Prometheus metrics format for performance monitoring.

## How Has This Been Tested?

`cargo run`, `make test`, and CI on forked master branch. Build fails on Windows platforms, possibly due to missing RuntimeMetrics.

### CODE REVIEW

**Comment by joseluisq** [comment]
@pl4nty some comments after a quick look.

> Build fails on Windows platforms, possibly due to missing RuntimeMetrics.

Yes, it fails for other targets as well. I think we can enable the SWS metrics feature only for Linux as a first stage.

About the Tokio runtime metrics.
It's fine to provide only _runtime_ metrics for now. So let's focus on that.
I also believe that since the Tokio runtime metrics feature is unstable, we could introduce the SWS as experimental like `--experimental-metrics` once we make it work to be usable.
The reason is that we don't know when Tokio runtime metrics will be stable, so until then we have to signal users the intention via an experimental flag.

I didn't test the PR yet but I will give it a detailed look in the next few days. But for now, the initialization in `server.rs` looks ok.
Please feel free to continue updating the draft or ping me if necessary.

**Comment by joseluisq** [comment]
Also, it would be great to add a few tests and share some output of how it looks like as well as how it goes when integrating with a server that consumes the metrics endpoint.

**Comment by pl4nty** [comment]
this is `tokio_total_local_schedule_count` from https://wufb-ds.tplant.com.au/metrics visualised in Grafana:

image

**Comment by pl4nty** [comment]
Not sure why FreeBSD CI is failing, [NetBSD is passing on GitHub Actions](https://github.com/pl4nty/static-web-server/actions/runs/7409900604/job/20161056204). Looks like I can disable it for just Windows and FreeBSD if needed

**Comment by pl4nty** [comment]
For tests, I was thinking a simple `assert!(std::str::from_utf8(&body).unwrap().contains("tokio_workers_count"))` to start. I'm not sure how to implement tests with `RequestHandler` though. Do you have any guidance on how to mock it?

File: src/handler.rs [hunk 1]
```diff
@@ -176,6 +184,24 @@ impl RequestHandler {
...
+                    encoder
+                        .encode(&prometheus::default_registry().gather(), &mut buffer)
+                        .unwrap();
+                    let data = String::from_utf8(buffer.clone()).unwrap();
+
+                    Body::from(data)
+                } else {
+                    Body::empty()
+                };
+                let resp = Response::new(body);
```
Comment (joseluisq) [inline src/handler.rs]: Add the proper `Content-Type` for the Prometheus metrics response.

File: src/handler.rs [hunk 2]
```diff
@@ -80,6 +82,8 @@ pub struct RequestHandlerOpts {
     pub ignore_hidden_files: bool,
     /// Health endpoint feature.
     pub health: bool,
+    /// Metrics endpoint feature.
```
Comment (joseluisq) [inline src/handler.rs]:
```suggestion
    /// Metrics endpoint feature (experimental).
```

File: src/server.rs [hunk 3]
```diff
@@ -292,6 +292,19 @@ impl Server {
         let health = general.health;
         server_info!("health endpoint: enabled={}", health);
 
+        // Metrics endpoint option
+        let metrics = general.metrics;
+        server_info!("experimental metrics endpoint: enabled={}", metrics);
```
Comment (joseluisq) [inline src/server.rs]:
```suggestion
        server_info!("metrics endpoint (experimental): enabled={}", metrics);
```

File: src/server.rs [hunk 4]
```diff
@@ -292,6 +292,19 @@ impl Server {
         let health = general.health;
         server_info!("health endpoint: enabled={}", health);
 
+        // Metrics endpoint option
```
Comment (joseluisq) [inline src/server.rs]:
```suggestion
        // Metrics endpoint option (experimental)
```

File: src/handler.rs [hunk 5]
```diff
@@ -176,6 +184,24 @@ impl RequestHandler {
                 return Ok(resp);
             }
 
+            // Metrics endpoint check
+            if metrics_request {
```
Comment (joseluisq) [inline src/handler.rs]: Note that I moved `use prometheus::Encoder;` down here.

```suggestion
            #[cfg(linux)]
            if metrics_request {
                use prometheus::Encoder;

```

File: src/server.rs [hunk 6]
```diff
@@ -292,6 +292,19 @@ impl Server {
...
+        server_info!("experimental metrics endpoint: enabled={}", metrics);
+        if metrics {
+            // not yet supported on Windows
+            #[cfg(not(windows))]
+            prometheus::default_registry()
+                .register(Box::new(
+                    tokio_metrics_collector::default_runtime_collector(),
+                ))
+                .unwrap();
+        }
```
Comment (joseluisq) [inline src/server.rs]:
```suggestion
        #[cfg(linux)]
        if metrics {
            prometheus::default_registry()
                .register(Box::new(
                    tokio_metrics_collector::default_runtime_collector(),
                ))
                .unwrap();
        }
```

File: src/server.rs [hunk 7]
```diff
@@ -332,6 +345,7 @@ impl Server {
                 ignore_hidden_files,
                 index_files,
                 health,
+                metrics,
```
Comment (joseluisq) [inline src/server.rs]:
```suggestion
                #[cfg(linux)]
                metrics,
```

File: src/settings/cli.rs [hunk 8]
```diff
@@ -400,6 +400,18 @@ pub struct General {
     /// This is especially useful with Kubernetes liveness and readiness probes.
     pub health: bool,
 
+    #[arg(
```
Comment (joseluisq) [inline src/settings/cli.rs]:
```suggestion
    #[cfg(linux)]
    #[arg(
```

File: src/settings/mod.rs [hunk 9]
```diff
@@ -147,6 +147,7 @@ impl Settings {
         let mut ignore_hidden_files = opts.ignore_hidden_files;
         let mut index_files = opts.index_files;
         let mut health = opts.health;
+        let mut metrics = opts.metrics;
```
Comment (joseluisq) [inline src/settings/mod.rs]:
```suggestion
        #[cfg(linux)]
        let mut metrics = opts.metrics;
```

File: src/settings/mod.rs [hunk 10]
```diff
@@ -292,6 +293,9 @@ impl Settings {
                 if let Some(v) = general.health {
                     health = v
                 }
+                if let Some(v) = general.metrics {
+                    metrics = v
+                }
```
Comment (joseluisq) [inline src/settings/mod.rs]:
```suggestion
                #[cfg(linux)]
                if let Some(v) = general.metrics {
                    metrics = v
                }
```

File: src/settings/mod.rs [hunk 11]
```diff
@@ -540,6 +544,7 @@ impl Settings {
                 ignore_hidden_files,
                 index_files,
                 health,
+                metrics,
```
Comment (joseluisq) [inline src/settings/mod.rs]:
```suggestion
                #[cfg(linux)]
                metrics,
```

File: src/handler.rs [hunk 12]
```diff
@@ -122,13 +126,17 @@ impl RequestHandler {
...
         let health = self.opts.health;
+        let metrics = self.opts.metrics;
         let index_files: Vec<&str> = self.opts.index_files.iter().map(|s| s.as_str()).collect();
 
         let mut cors_headers: Option<HeaderMap> = None;
 
         let health_request =
             health && uri_path == "/health" && (method.is_get() || method.is_head());
 
+        let metrics_request =
```
Comment (joseluisq) [inline src/handler.rs]:
```suggestion
        #[cfg(linux)]
        let metrics_request =
```

File: src/handler.rs [hunk 13]
```diff
@@ -176,6 +184,24 @@ impl RequestHandler {
...
+                    Body::from(data)
+                } else {
+                    Body::empty()
+                };
+                let resp = Response::new(body);
+                return Ok(resp);
+            }
+
             // Reject in case of incoming HTTP request method is not allowed
             if !method.is_allowed() {
```
Comment (joseluisq) [inline src/handler.rs]: When the changes above are done then move the whole "metrics endpoint check if block" to line 215. (before the CORS check)

File: src/handler.rs [hunk 14]
```diff
@@ -31,6 +31,8 @@ use crate::{
 #[cfg(feature = "directory-listing")]
 use crate::directory_listing::DirListFmt;
 
+use prometheus::Encoder;
```
Comment (joseluisq) [inline src/handler.rs]:
```suggestion
```

File: src/settings/file.rs [hunk 15]
```diff
@@ -237,6 +237,9 @@ pub struct General {
     /// Health endpoint feature.
     pub health: Option<bool>,
 
+    /// Metrics endpoint feature.
+    pub metrics: Option<bool>,
```
Comment (joseluisq) [inline src/settings/file.rs]:
```suggestion
    /// Metrics endpoint feature (experimental).
    #[cfg(linux)]
    pub metrics: Option<bool>,
```

**Review by joseluisq — CHANGES_REQUESTED** [review]
Some minor changes to make for now. When those are done then I will suggest a few more.

**Comment by joseluisq** [comment]
> Not sure why FreeBSD CI is failing, [NetBSD is passing on GitHub Actions](https://github.com/pl4nty/static-web-server/actions/runs/7409900604/job/20161056204). Looks like I can disable it for just Windows and FreeBSD if needed

It looks like it can not be found most likely because it is not supported yet or something. Let's focus on Linux only.

> For tests, I was thinking a simple `assert!(std::str::from_utf8(&body).unwrap().contains("tokio_workers_count"))` to start. I'm not sure how to implement tests with `RequestHandler` though. Do you have any guidance on how to mock it?

You can use this test below as a reference and create a new `tests/prometheus_metrics.rs` file. Also, you could add two tests one with the feature enabled and another disabled.

https://github.com/static-web-server/static-web-server/blob/8815762f56fe7d499b50c4f598b5e90520c09e2e/tests/dir_listing.rs#L181-L222

**Comment by joseluisq** [comment]
@pl4nty about testing, please forget my previous comment.
Recently, I published a testing utility and some test examples that you can use as a reference to test `RequestHandler`.

See https://github.com/static-web-server/static-web-server/pull/301/commits/76ab978c00899b8c51b8819b2505af62b5fecc30

**Comment by yonas** [comment]
I can build on FreeBSD by passing `RUSTFLAGS="--cfg tokio_unstable"` to `cargo build --release`.

I'm not sure why it's not picking up this setting from `.cargo/config.toml` like the other distros.

**Comment by joseluisq** [comment]
For some strange reason, GH closed the PR (or I did it unintentionally) when I was force-pushing changes on top and now I can not reopen it.

@pl4nty I will simply create a new PR with your changes on top.

**Comment by joseluisq** [comment]
Please let's continue the work on #307.

**Pull request closed without merging**